OPAL/br/src/main/scala/org/opalj/br/cfg/CFG.scala

```diff
@@ -722,6 +728,94 @@ case class CFG[I <: AnyRef, C <: CodeSequence[I]](
+        if (naturalLoops.isEmpty) {
+            val domTree = dominatorTree
+            // Execute a depth-first-search to find all back-edges
+            val start = startBlock.startPC
+            val seenNodes = ListBuffer[Int](start)
+            val toVisitStack = mutable.Stack[Int](successors(start).toArray: _*)
+            // backedges stores all back-edges in the form (from, to) (where to dominates from)
+            val backedges = ListBuffer[(Int, Int)]()
+            while (toVisitStack.nonEmpty) {
+                val from = toVisitStack.pop()
+                val to = successors(from).toArray
+                // Check for back-edges (exclude catch nodes here as this would detect loops where
+                // no actually are
+                to.filter { next ⇒
+                    val index = seenNodes.indexOf(next)
+                    val isCatchNode = catchNodes.exists(_.handlerPC == next)
+                    index > -1 && !isCatchNode && domTree.doesDominate(seenNodes(index), from)
+                }.foreach { destIndex ⇒
+                    // There are loops that have more than one edge leaving the loop; let x denote
+                    // the loop header and y1, y2 two edges that leave the loop with y1 happens
+                    // before y2; this method only saves one loop per loop header, thus y1 is
+                    // removed as it is still implicitly contained in the loop denoted by x to y2
+                    // (note that this does not apply for nested loops, they are kept)
+                    val hasDest = backedges.exists(_._2 == destIndex)
+                    var removedBackedge = false
+                    backedges.filter {
+                        case (oldTo: Int, oldFrom: Int) ⇒ oldFrom == destIndex && oldTo < from
+                    }.foreach { toRemove ⇒ removedBackedge = true; backedges -= toRemove }
+                    if (!hasDest || removedBackedge) {
+                        backedges.append((from, destIndex))
+                    }
+                }
+
+                seenNodes.append(from)
+                toVisitStack.pushAll(to.filter(!seenNodes.contains(_)))
+            }
+
+            // Finally, assemble the lists of loop elements
+            naturalLoops = Some(backedges.map { case (dest, root) ⇒ root.to(dest).toList }.toList)
+        }
+
+        naturalLoops.get
+    }
+
+    /**
+     * @return Returns the post dominator tree of this CFG.
+     *
+     * @see [[PostDominatorTree.apply]]
+     */
+    def postDominatorTree: PostDominatorTree = {
+        val exitNodes = basicBlocks.zipWithIndex.filter { next ⇒
+            next._1.successors.size == 1 && next._1.successors.head.isInstanceOf[ExitNode]
+        }.map(_._2)
+        PostDominatorTree(
+            if (exitNodes.length == 1) Some(exitNodes.head) else None,
+            i ⇒ exitNodes.contains(i),
+            // TODO: Pass an IntTrieSet if exitNodes contains more than one element
```

Comment:
Looks like these two TODOs should be addressed before merging.

Reply:
I'm unsure about the semantics of `additionalExitNodes` for a `PostDominatorTree`. I've now implemented a solution where `additionalExitNodes` are all nodes leading into an instance of type `ExitNode` *that is NOT* the `normalReturnNode`. However, i found some documentation about `additionalExitNodes` being used for escaping infinite loops, and i am not sure whether "normal" abnormal termination, like exceptions, belongs into the set of `additionalExitNodes` or not. Do you have any information or pointers for that?

Reply:
This seems to be used only in RecordCFG, line 967ff for handling infinite loops.
I wonder, though, if there is even a need for this whole method here, if RecordCFG already provides a post-dominator tree? Maybe we could just tap into that?

Reply:
Outdated since handling complex control structures is not relevant to this PR anymore.